### DISCUSSION TITLE
Ray tune with Yolo

### DISCUSSION THREAD

**Discussion by Coder-Vishali** [ultralytics/yolov5#10371] [Q&A]
I like to know if anyone have used ray tune hyperparameter tuning with YOLO models. If yes, please share me some pointers.

**Comment by glenn-jocher** [comment]
@Coder-Vishali not yet, but if you have luck with it please consider creating a tutorial for other users!

@AyushExel

**Reply by jifrah** [reply]
Any update on that?

**Comment by erminkev1** [comment]
Hello :) 

I have managed to make RayTune work with YOLOV5, by restructuring the code that is used to train the model. 
Basically, there were several considerations I had taken into account when making this work. I am quite sure there are other ways to make the tuning work, but this has worked out for me too. 

First, I considered the RayTune implementation in YOLOv8, specifically the function [_run_ray_tune()_](https://github.com/ultralytics/ultralytics/blob/main/ultralytics/utils/tuner.py#L9). You see that the function accepts, among a few others,  a _model_ and _train-args_ as its arguments. And inside the same function, at [this line](https://github.com/ultralytics/ultralytics/blob/main/ultralytics/utils/tuner.py#L91), you see that the function is defined which starts training of the model [ here  ](https://github.com/ultralytics/ultralytics/blob/main/ultralytics/utils/tuner.py#L91). So, this hinted me to restructure the _train,py_ function to make it work.

Therefore, secondly I looked at the _train.py_ function and I realized that inside of it, whenever it is called, a new model is instantiated, and then this model is subsequently trained. I thought that such approach would be a bit redundant for tuning so I purposefully took out this part from model training, and made it so that a model is created using another external function (using exactly the same code as inside train.py) 

As for the training arguments, instead of the argument parser, I changed the training function so that these arguments can be passed manually in my Python script, rather than inside a terminal. So, the training script in its most general way should be something like;

`def train(model, arg1, arg2, arg3 ....)` where _arg1, arg2, arg3 ..._ are all the different possible arguments used in training. Now these could be changed manually, and of course all of them had some default value, just like they do if you use input via terminal. 

So now that you have a function that makes a model, and a function that trains a model based on the input training arguments, you can use this in the _run_ray_tune quite easily. Following the principle of [v8 implementation](https://github.com/ultralytics/ultralytics/blob/main/ultralytics/utils/tuner.py#L91), instead of running `results = model_to_train.train(**config)`, it was something like `my_train_fuc(model, **config)` in my case. 

I hope you manage to make it work based off my explanation. I am still making the code a bit pretty, and soon I will share the code with the community.

Kind regards,
Ermin

**Reply by HabilKhanSuratee** [reply]
Hey, any chance could you share how you made this work? trying this out and struggling to do so.
